Bring CDEPS Inline capability to FV3

uturuncoglu opened the pull request:
<!-- INSTRUCTIONS: 
- PLEASE READ/FOLLOW THE DIRECTIONS IN EACH SECTION
- Complete the 'Commit Queue Requirements' below
- Please use github markup as much as possible (https://docs.github.com/en/get-started/writing-on-github)
- Please leave your PR in a draft state until all underlying work is completed.
-->
## Commit Queue Requirements:
<!--
- Please complete the items that follow this.
- Please "check off" completed items. Use [X] for a filled in checkbox or leave it [ ] for an empty checkbox
- Your PR will not be considered until all requirements are met.
- THIS IS YOUR RESPONSIBILITY
 -->
- [ ] Fill out all sections of this template.
- [ ] All sub component pull requests have been reviewed by their code managers.
- [ ] Run the full Intel+GNU RT suite (compared to current baselines) on either Hera/Derecho/Hercules
- [ ] Commit 'test_changes.list' from previous step
---
## Description:
<!--
Please provide a detailed verbose description of what this PR does
-->
This PR aims to bring CDEPS Inline capability to FV3 to allow updating desired fields in CCPP.

### Commit Message:
<!--
Please provide concise information for The UFS-WM and/or each sub-component:
Please delete what is not needed.
-->
```
* UFSWM - 
  * AQM - 
  * CDEPS - 
  * CICE - 
  * CMEPS - 
  * CMakeModules - 
  * FV3 - 
    * ccpp-physics - 
    * atmos_cubed_sphere - 
  * GOCART - 
  * HYCOM - 
  * MOM6 - 
  * NOAHMP - 
  * WW3 - 
  * fire_behavior
  * stochastic_physics - 
```

### Priority:
<!--
Please provide the priority you would prefer this pull request to have.
* Critical Bugfix: Model is wrong.
* High: Time-sensitive project.
* Normal.
Please delete the ones that are not applicable
-->
* Normal

## Git Tracking
### UFSWM:
<!--
Please add the UFS-WM github issue here if there is one
Please delete the one that is not applicable.
-->
* None

### Sub component Pull Requests:
<!--
Please provide a list of sub-components involved with this pull request.
Please provide links to the sub-component pull requests as shown below.
Please delete what is not needed.
Example:
* FV3: NOAA-EMC/fv3atm#734
  * ccpp-physics: ufs-community/ccpp-physics#33
* WW3: NOAA-EMC/WW3#321
-->
* AQM:
* CDEPS:
* CICE:
* CMEPS:
* CMakeModules:
* FV3: https://github.com/NOAA-EMC/fv3atm/pull/988
  * ccpp-physics: https://github.com/ufs-community/ccpp-physics/pull/298
  * atmos_cubed_sphere:
* GOCART:
* HYCOM:
* MOM6:
* NOAHMP:
* WW3:
* fire_behavior:
* stochastic_physics:
* None

### UFSWM Blocking Dependencies:
<!--
If there are any UFSWM PR's that are needed to be completed before this one, please add links
to them here
Please delete what is not needed.
-->
* None

### Documentation:
<!--
Indicate what documentation, if any, is needed for this PR and who will add it (if applicable).
Please delete what is not needed.
-->
* This PR requires a documentation update, and the WM User's Guide has been updated based on the changes in this PR.
* This PR requires a documentation update, and a WM issue has been opened to track the need for a documentation update; a person responsible for submitting the update has been assigned to the issue (link issue). 
* No documentation update is required for this PR (please explain).

---
## Changes
### Regression Test Changes (Please commit test_changes.list):
<!--
Please let us know if this PR creates new baselines, changes baselines or not.
Please delete what is not needed.
Please make sure you have properly submitted test_changes.list
-->
* PR Adds New Tests/Baselines.

### Input data Changes:
<!--
If there are any changes to input-data for a test, please provide information here.
Please delete what is not needed.
-->
* New input data.

### Library Changes/Upgrades:
<!-- Library updates take time. Please provide library and version information here.
** SPECIAL INSTRUCTIONS **
If this PR needs updates to libraries please make sure to accomplish the following tasks:
- Create separate issue in (https://github.com/JCSDA/spack-stack) asking for update to library. Include library name, library version.
- Add issue link from JCSDA/spack-stack following this item <!-- for example: "* JCSDA/spack-stack#1757"

Please delete what is not needed.
-->
* No Updates
  
---
<!-- STOP!!! THE FOLLOWING IS FOR CODE MANAGERS ONLY. PLEASE DO NOT FILL OUT -->
## Testing Log:
- RDHPCS
  - [ ] Hera
  - [ ] Orion
  - [ ] Hercules
  - [ ] GaeaC6
  - [ ] Derecho
  - [ ] Ursa
- WCOSS2
  - [ ] Dogwood/Cactus
  - [ ] Acorn
- [ ] CI
- [ ] opnReqTest (complete task if unnecessary)

NickSzapiro-NOAA commented:
Thank you @uturuncoglu!

I am still catching up but I do see some reshape calls that may be important for GFS_coupling --> CCPP variables:
https://github.com/search?q=repo%3ANOAA-EMC%2Ffv3atm%20reshape&type=code

If I understand, maybe it is worth checking what lat,lon from FV3 mesh look like if passed over to CCPP in same way

uturuncoglu commented:
@NickSzapiro-NOAA I think the issue is solved when I make change in the standard names after @DusanJovic-NOAA suggestion. At this point I am not sure what is the difference between `horizontal_dimension` vs. `horizontal_loop_extent` in the definition and how was it working in the first place before CCPP RP in the past  but it looks it fixed the issue. See following commit - https://github.com/NOAA-EMC/fv3atm/pull/988#issuecomment-3066112570. I am running model with/without inline at this point to be sure it is working. The fields are in CCPP side looks good at this point. I'll add couple of plots to the PR.

uturuncoglu commented:
@NickSzapiro-NOAA @junwang-noaa @DeniseWorthen It seems issue is fixed. @DusanJovic-NOAA Thanks again for your great help. Please see the latest comment in https://github.com/NOAA-EMC/fv3atm/pull/988. Let me know if you want me to change those PRs ready to view.

uturuncoglu commented:
Just let me know if you also want me to include a new RT that demonstrates the capability.

uturuncoglu commented:
@junwang-noaa @DeniseWorthen @NickSzapiro-NOAA JFYI, I'll add a new RT (regional - RRFS_CONUScompact_13km, updates GL SST and ice conditions) to demonstrate the capability and this PR will be ready to review and merge. The test is created with SRW app.

uturuncoglu commented:
I think I could modify existing `conus13km_control` and enable inline capability in it. So, new test would be `conus13km_control_inline`.

uturuncoglu commented:
@junwang-noaa @DeniseWorthen @NickSzapiro-NOAA Okay. I added new test - `conus13km_gl_inline`. This is little bit different configuration then other `conus13km`in terms of `input.nml` and also the simulation period (I need to be consistent with the FVCOM forecast used in the CDEPS Inline). At this point, this test is getting `input.nml` and stream.config configuration files from input directory - but if you really need I need to spend more time to generate those from template. Since there is no any test that uses CDEPS Inline, maybe keeping like this would be okay. Let me know what you think. I also updated PR description to include new input folder and add new RT information and make this PR ready to review.

uturuncoglu commented:
@NickSzapiro-NOAA I am not sure what is wrong on Derecho but RT system marks the new test failed even if it is successful.

```
++ qstat 2106516 -x -f
++ grep Exit_status
++ awk '{print $3}'
+ exit_status=
+ set -e
+ [[ '' != 0 ]]
+ status=FAILED
```

It seems that rt.sh checks the job with `qstat 2106516 -x -f | grep Exit_status | awk '{print $3}'` command. In my case, if I run the command manually, it returns 0 but in rt.sh it returns empty. Anyway, since I did not make any change in the rt system except adding new test, I think that is fine. My source is in `/glade/work/turuncu/GL/ufs-weather-model_test` if you want to check logs.

NickSzapiro-NOAA commented on tests/rt.conf:
```diff
@@ -229,6 +229,9 @@ RUN | conus13km_control                                 | - noaacloud
 RUN | conus13km_restart_mismatch                        | - noaacloud                          | baseline | conus13km_control
 
+COMPILE | rrfs_dyn32_phy32_faster | intel | -DAPP=ATM -DFASTER=ON -DCCPP_SUITES=FV3_HRRR -D32BIT=ON -DCCPP_32BIT=ON -DCDEPS=ON -DCDEPS_INLINE=ON | - noaacloud  | fv3 |
+RUN | conus13km_gl_inline                               | - noaacloud                          |          |
```
conus13km_gl_inline needs its own baseline

uturuncoglu replied:
Okay. Let me add it. I created a baseline to check but I did not add baseline indicator to the rt.conf. Thanks for your help.

uturuncoglu replied:
@NickSzapiro-NOAA Still getting same error - /glade/work/turuncu/GL/ufs-weather-model_test/tests/gs/log_derecho/run_conus13km_gl_inline_intel.log. I'll try to one of the other test to see what happens.

NickSzapiro-NOAA commented:
I see what you're saying @uturuncoglu . Could there be some lag for qstat between job_running=false and Exit_status being assigned?

uturuncoglu commented:
@NickSzapiro-NOAA Not sure but maybe I could try to run on Hercules to see if I have similar issue also in there.

uturuncoglu commented:
@NickSzapiro-NOAA I am also getting similar error from UFS WM head of develop. For example, I tested with `datm_cdeps_control_gefs_intel` and it is failing with same error. I am not sure it is a system issue or something in my account. I'll try double check but if you have account on Derecho could you try to run of of the RT to see if there is an issue.

NickSzapiro-NOAA commented:
Yes, there is some system issue with derecho (maybe after last week's maintenance). It is being skipped in commit queue RTs right now

uturuncoglu commented:
@NickSzapiro-NOAA Okay. good to know. Anyway, I am not sure about the timeline in the commit queue but this PR ready to merge. I'll keep syncing. Let me know if you need anything from my end.

gspetro-NOAA commented:
@uturuncoglu Could you sync up your branch and push a log file showing your testing? We are going to start requiring developers to push a log file as part of our pre-testing checks (preference is an Ursa one, but Hercules or Derecho will do). If this PR is still ready to go, then I will raise it at the CM meeting as one to schedule.

uturuncoglu commented:
@gspetro-NOAA Sure. I'll sync it soon. I have access only to Hercules. Let me know which command do I need to use to generate those logs.

gspetro-NOAA commented:
@uturuncoglu You could run `./rt.sh -a <account> -e &`, which will generate a log called `ufs-weather-model/tests/logs/RegressionTest_hercules.log`. You would push that log file and also `test_changes.list`, which will show which tests have changed.

uturuncoglu commented:
@gspetro-NOAA Okay. It was challenging since FV3 renamed as UFSATM now but synced now. I am starting to test and push the log files when it is done.

uturuncoglu commented:
@gspetro-NOAA I have just push the log file. RTs are fine on Hercules.

uturuncoglu commented:
@gspetro-NOAA BTW, I need to run the new test on Derecho (input files are staged in there) to be sure it is still working fine.

uturuncoglu commented:
@gspetro-NOAA I am getting following error when I try to compile new RT (conus13km_gl_inline) on Derecho,

```
Lmod has detected the following error: These module(s) or extension(s) exist
but cannot be loaded as requested: "scotch/7.0.4"
   Try: "module spider scotch/7.0.4" to see how to load the module(s).
```

This was working fine in the past. Let me know what you think? I could also copy input files to Hercules and try over there.

gspetro-NOAA commented:
@uturuncoglu At the moment, we are not running RTs on Derecho until PR #2863 goes through to fix a number of issues. If you can test on Hercules, that will speed things up. Otherwise, we may have to view #2863 as a dependency. @jkbk2004 Anything to add?

uturuncoglu commented:
@gspetro-NOAA Okay. I'll test on Hercules and update you. Copying required files now.

uturuncoglu commented:
@gspetro-NOAA Okay. The case is working fine on Hercules. The only issue that I noticed. The RT system assigning wrong number of processor to component (assigning 85 core to FV3 but layout was 12x12 in input.nml). So, I had to manually edit processor counts in config files. This was working fine on Derecho. I wonder if something changed in UFS WM model side. Any idea?

uturuncoglu commented:
@gspetro-NOAA Okay. I found the issue and will be fix soon. I was getting input.nml from INPUT directory to make it simple but I might need to define a template for it.

uturuncoglu commented:
@gspetro-NOAA I also fixed the test. So, it is ready in my end.

gspetro-NOAA commented:
Great! Could you push your logs from your test run?

uturuncoglu commented:
@gspetro-NOAA I have already pushed the test change list. Let me know if you want mw to push the logs too.

gspetro-NOAA commented:
@uturuncoglu Yes, ideally developers should push the logs, too, unless the changes are simple documentation/CI changes that don't affect source code. If you can do that and get the UFSATM subcomponent PR approved, then we can schedule this PR for the Commit Queue.

uturuncoglu commented:
@gspetro-NOAA Okay. That makes sense. Let me know which log file do I need to push. I could push `RegressionTests_derecho.log` but I think that would be generated by automated testing. Maybe I am wrong.

gspetro-NOAA commented:
@uturuncoglu Since you did your most recent testing on Hercules, you can push `tests/logs/RegressionTests_hercules.log`.

gspetro-NOAA commented:
@uturuncoglu Just looking for a status update on this because I haven't seen any updates to the logs/test_changes.list.

uturuncoglu commented:
@gspetro-NOAA Let me push test change list and also `tests/logs/RegressionTests_hercules.log`.

uturuncoglu commented:
@gspetro-NOAA Okay. I rerun all tests on Hercules and here is the failed tests,

```
fail_test_conus13km_gl_inline_intel
fail_test_hafs_regional_atm_wav_intel
fail_test_regional_atmaq_debug_intel
```

I could not see anything particular in the log files and it seems `hafs_regional_atm_wav_intel` is run without any issue and `regional_atmaq_debug` is failed because of the wall time limit. Any idea? I think I could not push the log file since it will also show these two failed runs but if you would like, I could push them too. BTW, my run directory is in `/work2/noaa/stmp/tufuk/stmp/tufuk/FV3_RT/rt_643896`

gspetro-NOAA commented:
Ok, thanks! If you can sync w/WM develop branch, I can finalize the test_changes.list stuff for you, and we can get this into the commit queue. :)

uturuncoglu commented:
@gspetro-NOAA Thanks. I have just sync the model. Let me know if you have any issues.

gspetro-NOAA commented:
@uturuncoglu I'm running tests on Ursa, and everything is failing in compile. I haven't checked every `err` file, but they generally seem to have a message like this: 
```
CMake Error at UFSATM/ccpp/CMakeLists.txt:74 (message):
  An error occured while running ccpp_prebuild.py, check
  /scratch4/NAGAPE/epic/Gillian.Petro/stmp/RT_RUNDIRS/Gillian.Petro/FV3_RT/rt_1014888/compile_s2sw_intel/build_fv3_s2sw_intel/UFSATM/ccpp/ccpp_prebuild.{out,err}
```
This is the same message in the failing Spack test below.

uturuncoglu commented:
@gspetro-NOAA Let me check. Maybe it is introduced when I sync with UFSATM yesterday.

uturuncoglu commented:
@gspetro-NOAA There was an issue when syncing UFSATM. I fixed it now. Testing in my end. I hope GitHub Action will pass now.

uturuncoglu commented:
@gspetro-NOAA `control_p8` is passing in my end. I'll also run a coupled configuration to be sure.

uturuncoglu commented:
@gspetro-NOAA `cpld_control_p8` is also passing in my end.

gspetro-NOAA commented:
Ok, I uploaded the Ursa log and the `test_changes.list` file. Everything passes on Ursa except for the new test. Basically, we just need to stage the new input data for the new test on all systems. However, Denise's PR (#2835) will update to a new input-data-directory, so I think I will hold off on staging this data until her PR gets in (should be scheduled soon--hopefully early next week). That way I don't need to update two directories per machine w/your data.

gspetro-NOAA commented:
@uturuncoglu On your side, all we need is for you to fill out the "Commit Message" section of the PR template. We will take care of the input data and then request the UFSATM review.

uturuncoglu commented:
@gspetro-NOAA Okay. I updated the part related with the commit message. Let me know if you need anything in my end.

gspetro-NOAA commented:
@uturuncoglu Now that Denise's PR with the new input data directory has been merged, I can work on getting your data propagated to all systems, and then we can add this PR to the commit queue. 
I just want to clarify -- are there any changes to existing input data? It looks like you are only adding new data for a new test, but if there are any changes to existing data, we'll need to create a new input data directory.

gspetro-NOAA commented:
@uturuncoglu Just wanted to follow up on the input data question above so that we can move forward on this PR. Are there any changes to existing input data w/this PR?

uturuncoglu commented:
@gspetro-NOAA There is no any change in the input files. We are working on some changes in CCPP physics but it is not related with input. So, that is fine for me to push input files to all platforms.

gspetro-NOAA commented:
@dpsarmie @NickSzapiro-NOAA One of you suggested adding the input data for Ufuk's new `conus13km_inline_gl` case in a CDEPS directory. It seems (if I've understood correctly), that there's an active FV3 component coupled with CDEPS (DOCN?) We might want to name the directory in a way that shows that, e.g., `FV3_input_data_cdeps_inline`. Does this work? Or can you suggest a better directory name?

NickSzapiro-NOAA commented:
Sure! I think was wrong about where to put the new input data. A new FV3_input_data_cdeps_inline seems to fit in with all the other FV3_input_data* directories. Does that work @uturuncoglu ?

fyi, the new test is ATM and doesn't couple with a CDEPS data component. Rather, ATM calls CDEPS source code "inline" to read (Great Lakes) surface data from file and passes those conditions down into CCPP physics.

DeniseWorthen commented:
It looks to me that the new ``rrfs_cdeps_inline_run.IN`` expects the data to be in a very specific place:

```
cp -r @[INPUTDATA_ROOT]/FV3_input_data_conus13km_inline_GL/INPUT .
cp -r @[INPUTDATA_ROOT]/FV3_input_data_conus13km_inline_GL/INPUT_DATA .
cp    @[INPUTDATA_ROOT]/FV3_input_data_conus13km_inline_GL/stream.config .
cp    @[INPUTDATA_ROOT]/FV3_aeroclim/* .


## Thompson MP lookup tables - copy standard and SIONlib tables of precomputed tables
if [ $IMP_PHYSICS = 8 ]; then
  cp    @[INPUTDATA_ROOT]/FV3_fix/qr_acr_qsV2.dat .
  cp    @[INPUTDATA_ROOT]/FV3_fix/qr_acr_qgV2.dat .
  cp    @[INPUTDATA_ROOT]/FV3_fix/freezeH2O.dat .
  cp    @[INPUTDATA_ROOT]/FV3_fix/CCN_ACTIVATE.BIN .
fi
```

uturuncoglu commented:
@NickSzapiro-NOAA @gspetro-NOAA @DeniseWorthen Any directory will work for me. If you change the name of the exiting one just let me know and I could update `rrfs_cdeps_inline_run.IN`

gspetro-NOAA commented:
> * /glade/work/turuncu/GL/RT/NEMSfv3gfs/input-data-20250507/FV3_input_data_conus13km_inline_GL/INPUT

@uturuncoglu Thanks! I've put the data under `FV3_input_data_cdeps_inline` on all systems except Gaea, where I'm having role account issues. Will see if I can resolve that or get someone else to stage data there. In the meantime, if you can update the paths in the PR, that would be great. I think we're almost ready to go!

gspetro-NOAA commented:
@uturuncoglu Data is now available on Gaea C6 (so all RDHPCS), so all we need from you is your path update, and I can run some final tests before we schedule the PR. 
@BrianCurtis-NOAA I think you'll need to add Ufuk's data on WCOSS2 and Acorn. We decided on `FV3_input_data_cdeps_inline` for the directory name.

gspetro-NOAA commented:
@uturuncoglu We're just waiting on your path updates to `rrfs_cdeps_inline_run.IN`. The data was placed in `@[INPUTDATA_ROOT]/FV3_input_data_cdeps_inline` on all systems.

uturuncoglu commented:
@gspetro-NOAA Let me check with the science team. There was some issue in the CCPP side and I am waiting confirmation from the team. @grantfirl was helping us about it and created a PR to my branch to fix it. @grantfirl do you have any update about the it. I know you created a PR to my branch but I was waiting to have all issues resolved. Anyway, let me ask the team again. @gspetro-NOAA I'll update you soon.

uturuncoglu commented:
@gspetro-NOAA Sorry for delay but we were waiting confirmation from the science team. Anyway, I merged @grantfirl mods for CCPP physics and sync all the components as well as model. I also update the input directory for the specific test but I realized only part of input files is in `/glade/derecho/scratch/epicufsrt/ufs-weather-model/RT/NEMSfv3gfs/input-data-20251015/FV3_input_data_cdeps_inline` on Derecho. For example, there is no `INPUT_DATA` folder and `stream.config` file over there. If you don't mind could you double check? Then, I could try to run the same test again in my end. I think that was my fault since I put only INPUT folder to input files section. `INPUT_DATA` folder and `stream.config` also needs to be copied. All the files are in `/glade/work/turuncu/GL/RT/NEMSfv3gfs/input-data-20250507/FV3_input_data_conus13km_inline_GL` on Derecho. Sorry about my fault, I also fixed in the description.

gspetro-NOAA commented:
@uturuncoglu I have added the missing data to the EPIC role locations on Derecho and Ursa, so it should work now. I will stage it on the other RDHPCS shortly. 
@BrianCurtis-NOAA I have restaged the full data set on Ursa at `/scratch4/NAGAPE/epic/role-epic/UFS-WM_RT/NEMSfv3gfs/input-data-20251015/FV3_input_data_cdeps_inline`, so it should be ready to move to WCOSS2/Acorn.

UPDATE: I have finished staging data on supported Level 1 RDHPCS.

uturuncoglu commented:
@gspetro-NOAA Thanks for your help. I think there is still issue with the folder. It looks like `/glade/derecho/scratch/epicufsrt/ufs-weather-model/RT/NEMSfv3gfs/input-data-20251015/FV3_input_data_cdeps_inline/FV3_input_data_cdeps_inline/FV3_input_data_conus13km_inline_GL/`but the content of the `/glade/derecho/scratch/epicufsrt/ufs-weather-model/RT/NEMSfv3gfs/input-data-20251015/FV3_input_data_cdeps_inline/FV3_input_data_cdeps_inline/FV3_input_data_conus13km_inline_GL/` folder needs to be in `/glade/derecho/scratch/epicufsrt/ufs-weather-model/RT/NEMSfv3gfs/input-data-20251015/FV3_input_data_cdeps_inline`. So, the final input folder need to have only `INPUT`,  `INPUT_DATA`  and `stream.config`.

gspetro-NOAA commented:
@uturuncoglu I think I fixed it! I checked all the other systems, and it's set up how it's supposed to be; Derecho was the first one I did, so I think I didn't realize there was a problem when I moved that data. I only realized when it was already on other systems. On Hercules, data is at: `/work2/noaa/epic/hercules/UFS-WM_RT/NEMSfv3gfs/input-data-20251015/FV3_input_data_cdeps_inline`. You could try testing there if you want, too.

uturuncoglu commented:
@gspetro-NOAA Thanks all of your help. It looks fine on Derecho now. I am testing now and update you soon about it.

uturuncoglu commented:
@gspetro-NOAA I am having hard time to find `/glade/derecho/scratch/epicufsrt/ufs-weather-model/RT//NEMSfv3gfs/develop-20251223/` on Derecho. So, maybe Derecho is not the best place to test at this point. As I know there were some issue with spack-stack over there. Let me try on Hercules.

DeniseWorthen commented:
@uturuncoglu We haven't been running RTs on Derecho for quite some time, so there is not a current baseline to test against. Running on Hercules is your best bet.

gspetro-NOAA commented:
> @gspetro-NOAA I am having hard time to find `/glade/derecho/scratch/epicufsrt/ufs-weather-model/RT//NEMSfv3gfs/develop-20251223/` on Derecho. So, maybe Derecho is not the best place to test at this point. As I know there were some issue with spack-stack over there. Let me try on Hercules.

Yes, Denise is right--Hercules is best for now; there's still testing happening on the PR to reenable RTs on Derecho. 🙁

uturuncoglu commented:
@gspetro-NOAA Okay. It is working on Hercules. The only missing file is `/glade/work/turuncu/GL/RT/NEMSfv3gfs/input-data-20250507/FV3_input_data_conus13km_inline_GL/INPUT_DATA/mask.nc` this also needs to be placed under `/glade/derecho/scratch/epicufsrt/ufs-weather-model/RT/NEMSfv3gfs/input-data-20251015/FV3_input_data_cdeps_inline/INPUT_DATA/mask.nc`. Sorry about extra work.

gspetro-NOAA commented:
> @gspetro-NOAA Okay. It is working on Hercules. The only missing file is `/glade/work/turuncu/GL/RT/NEMSfv3gfs/input-data-20250507/FV3_input_data_conus13km_inline_GL/INPUT_DATA/mask.nc` this also needs to be placed under `/glade/derecho/scratch/epicufsrt/ufs-weather-model/RT/NEMSfv3gfs/input-data-20251015/FV3_input_data_cdeps_inline/INPUT_DATA/mask.nc`. Sorry about extra work.

Ok, I've moved that file onto the other RDHPCS. @BrianCurtis-NOAA will need to stage it on WCOSS2/Acorn when he's back from PTO.

gspetro-NOAA commented:
@uturuncoglu Is this PR waiting on anything more, or do you think it's ready for review/scheduling?

uturuncoglu commented:
@gspetro-NOAA Thanks for checking. I think this ready to go.

gspetro-NOAA commented:
@uturuncoglu I ran the RT suite on Ursa and Hercules, and most tests are failing now on both platforms... 😬  Currently looking into why. It mostly seems to be comparison failures, but I thought you were just adding two tests, not changing current baselines for existing tests. Some tests also failed to compile. 
On Hercules, my run_dir is: `/work2/noaa/epic/gpetro/hercules/RTs/ufs-wm/stmp/gpetro/FV3_RT/rt_3207276` and my WM clone is here: `/work2/noaa/epic/gpetro/hercules/RTs/ufs-wm/pretests/2807/tests`.

gspetro-NOAA commented:
For the compile failures, I'm seeing, for example: 
```
CMake Error at CMEPS-interface/CMakeLists.txt:142 (message):
  CCPP prebuild step failed, check
  /work2/noaa/epic/gpetro/hercules/RTs/ufs-wm/stmp/gpetro/FV3_RT/rt_3207276/compile_s2s_aoflux_intel/build_fv3_s2s_aoflux_intel/CMEPS-interface/ccpp_prebuild.{out,err}
```

uturuncoglu commented:
@gspetro-NOAA thanks for checking. in my end it was working fine and building. I could try to debug `compile_s2s_aoflux_intel` test to fix the issue. For answer change I am not sure. Let me work on it.
@grantfirl I wonder if you run any regular coupled RT after made changes in the CCPP level. Do you have any idea why they are failing in the baseline check. I think if we placed the controls in a correct way, there won't be any answer change in the configurations that do not use cdeps inline. So, we might forget to put some control into ccpp physics side to prevent it. Let me know what you think?

grantfirl commented:
@uturuncoglu Since the interface for GFS_surface_composites_post.F90 was changed (which is used in a CMEPS CCPP SDF), there will need to be changes to CMEPS to pass compilation. Do you want me to do this? Should be easy for me.

uturuncoglu commented:
@grantfirl Sure. If you help us about that that would be great. Do you think that that is the only issue. That might explain the issue with exchange grid test but other tests also showing answer change.

grantfirl commented:
> @grantfirl Sure. If you help us about that that would be great. Do you think that that is the only issue. That might explain the issue with exchange grid test but other tests also showing answer change.

The first thing that I would check would be that the latest develop code is merged in to the PR branches, which is why I requested that on ccpp-physics and ufsatm. If any recently-merged PRs changed results (that weren't included in the PR branches), that could be a culprit.

uturuncoglu commented:
@grantfirl Synced now (incl. ufsatm and ccpp/physics). Thanks for your help.

grantfirl commented:
> @grantfirl Synced now (incl. ufsatm and ccpp/physics). Thanks for your help.

@uturuncoglu OK, I have a CMEPS branch to test with this. I'll run RTs on Ursa to test the CMEPS branch and the re-synced branches together.

grantfirl commented:
@uturuncoglu RTs running now. I'll let you know how it turns out. Assuming the CMEPS branch works to fix the compilation failure, I'll start a PR in that repo. Then, you'll need to change .gitmodules in ufs-weather-model to point to the new CMEPS PR branch and `git add` it as well.

grantfirl commented:
@uturuncoglu The CMEPS branch works to fix the compilation issue, but I too am seeing a bunch of RT failures that shouldn't happen. I'll investigate this morning and see if I can find a culprit with the ccpp-physics changes. It's frustrating!

The CMEPS PR now exists: https://github.com/NOAA-EMC/CMEPS/pull/159

Please change .gitmodules to point to the CMEPS PR branch on my fork and `git add` the CMEPS submodule. The hash should be https://github.com/NOAA-EMC/CMEPS/pull/159/commits/f089e2842d5eebe15420510675b645c159fda324

uturuncoglu commented:
@gspetro-NOAA Thanks for your help. Let me update my branch and point CMEPS PR. I'll also update PR description.

uturuncoglu commented:
@gspetro-NOAA It is done. Let me know if you need help.

grantfirl commented:
@uturuncoglu Please take a look at the question that @NickSzapiro-NOAA brought up related to the CDEPS_INLINE compile option. https://github.com/NOAA-EMC/CMEPS/pull/159

uturuncoglu commented:
@gspetro-NOAA @NickSzapiro-NOAA The CDEPS_INLINE that I implemented for HAFS is different from the implementation that we are using for UFSATM. It seems that if you activate CDEPS_INLINE, it will activate both CMEPS and FV3 inline capability. But, CMEPS inline also needs a special phase in the run sequence and UFSATM inline also needs additional namelist option in input.nml. So, I do not think they will interface each other. Let me know if it is not clear and needs more explanation.

grantfirl commented:
@uturuncoglu @gspetro-NOAA I've found a bug leading to the spurious RT failures. I'm rerunning the entire RT suite with the fix to see if all RTs are fixed or just a subset.

uturuncoglu commented:
@grantfirl Thanks for your help. That is great. Let me know if you need anything from my end.

grantfirl commented:
> @grantfirl Thanks for your help. That is great. Let me know if you need anything from my end.

@uturuncoglu Please merge https://github.com/uturuncoglu/ccpp-physics/pull/3 into your PR branch of ccpp-physics and then update ufsatm and ufs-weather-model with the updated commit hashes. It's a minor bugfix that addresses all failed RTs except for the new test that is expected to fail.

I did see a test timeout of rap_control_dyn64_phy32_gnu and a write failure in cpld_control_p8_lnd_intel, but those should not be related to this PR.

uturuncoglu commented:
@grantfirl Sure. Thanks for your help.

uturuncoglu commented:
@gspetro-NOAA Okay. I merged your PR and sync the model.

gspetro-NOAA commented:
> @gspetro-NOAA Okay. I merged your PR and sync the model.

@uturuncoglu It looks like there are still conflicts in UFSATM.

uturuncoglu commented:
@gspetro-NOAA It is strange. I am not seeing any issue in the UFSATM level PR. Any idea?

grantfirl commented:
@uturuncoglu @gspetro-NOAA The only conflict that I'm seeing is that the ufs-weather-model PR branch needs to pull in the top of develop again. The UFSATM hash (https://github.com/NOAA-EMC/ufsatm/pull/988/commits/911026847b4b3c028c53f2e09642f3e1c7f83f43) looks OK.

uturuncoglu commented:
@grantfirl @gspetro-NOAA Okay. I have just synced WW3 and also model itself.

gspetro-NOAA commented:
> @grantfirl @gspetro-NOAA Okay. I have just synced WW3 and also model itself.

Thanks! Looks good. It's possible I hadn't refreshed the page. 🙄 Regardless, I'll go ahead and run some pretests on our end, and then hopefully we can get this scheduled.

gspetro-NOAA commented:
RTs passed for me on Hercules. Now we just need the subcomponent PRs to get reviewed (CDEPS and UFSATM), and we can schedule this PR.

gspetro-NOAA commented:
@uturuncoglu Could you please sync with develop? We'd like to begin processing this PR today. :)

uturuncoglu commented:
@grantfirl If you don't mind could you sync your CMEPS branch? We are pointing your fork at this point. Then, I'll update in the model level.

grantfirl commented:
> * add CDEPS inline data for GFS_surface_composites_pre/post NOAA-EMC/CMEPS#159

@uturuncoglu Done. Hash is https://github.com/NOAA-EMC/CMEPS/pull/159/commits/1199ba6615fda3798888373eaa6b520664d1fd03

uturuncoglu commented:
@grantfirl Thanks.
@gspetro-NOAA I have just sync the components and also model itself. Thanks for your help.

gspetro-NOAA commented:
@uturuncoglu In the CMEPS section of the PR template, could you add a commit message there?

uturuncoglu commented:
@gspetro-NOAA Done

gspetro-NOAA commented:
I am getting a persistent OOM error for `control_p8_ugwpv1_tempo_aerosol_hail_intel` on Ursa only. Here is my latest run_dir: `/scratch4/NAGAPE/epic/Gillian.Petro/stmp/RT_RUNDIRS/Gillian.Petro/FV3_RT/rt_115204`.
Some of the ORTs also failed. Rerunning and will check in the morning.

gspetro-NOAA commented:
After recloning and running the failing test on Ursa, it passed. 🎉 

UPDATE: ORTs also passed on rerun.

BrianCurtis-NOAA commented:
I think I typo'd my previous post, here's the hash

CMEPS Hash: https://github.com/NOAA-EMC/CMEPS/commit/4bb06b3f4834142b8ae4e9a6ab8a6aa7b1e098a7

Please revert .gitmodules for CMEPS and update the CMEPS hash to the one above.

BrianCurtis-NOAA commented:
UFSATM now at https://github.com/NOAA-EMC/ufsatm/commit/35e9c994e480bc6e16f67ca87db96f354f749c5e 

please revert those .gitmodules and update UFSATM hash to that above.

BrianCurtis-NOAA commented:
@uturuncoglu please make the changes mentioned previously. Thanks!

uturuncoglu commented:
@BrianCurtis-NOAA Sorry for the delay. I was in a meeting. I updated pointer for both UFSATM and CMEPS.

NickSzapiro-NOAA commented:
Thanks for all your efforts to get this committed @uturuncoglu

uturuncoglu commented:
@NickSzapiro-NOAA @gspetro-NOAA @BrianCurtis-NOAA Thanks all of you for your help. It was little bit bumpy PR. Sorry about it. I also want to have a spacial thank to @grantfirl for his contribution in CCPP/physics end.
